File: CodingTest/Python/solution10.py
# https://school.programmers.co.kr/learn/courses/30/lessons/42626
def solution(scoville, K):
    answer = 0
    import heapq
    heapq.heapify(scoville)
    while True:
        if len(scoville) <= 1 and scoville[0] < K:
            answer = -1
            break
        if scoville:
            first = heapq.heappop(scoville)
            if first >= K:
                break
        if scoville:
            second = heapq.heappop(scoville)
        new_food = first + second * 2
        heapq.heappush(scoville, new_food)
        answer += 1
    return answer

File: CodingTest/Java/Solution10.java
// https://school.programmers.co.kr/learn/courses/30/lessons/12911
package CodingTest.Java;

public class Solution10 {
    public int solution(int n) {
        int answer = 0;
        String binN = Integer.toBinaryString(n);
        int nn = n + (int) Math.pow(2, binN.length() - binN.lastIndexOf("1") - 1);
        String binNN = Integer.toBinaryString(nn);
        int nCnt = binN.length() - binN.replace("1", "").length();
        int nnCnt = binNN.length() - binNN.replace("1", "").length();

        int diff = nCnt - nnCnt;
        
        answer = nn + (int) Math.pow(2, diff) - 1;

        return answer;
    }
}


